File: bindings/kotlin/src/main/kotlin/com/hyperpolymath/proven/SafeChecksum.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

package com.hyperpolymath.proven

import kotlin.math.abs

/**
 * Checksum and hash utilities.
 */
object SafeChecksum {
    // CRC-32 lookup table
    private val CRC32_TABLE = IntArray(256) { i ->
        var crc = i
        repeat(8) {
            crc = if (crc and 1 != 0) (crc ushr 1) xor 0xEDB88320.toInt() else crc ushr 1
        }
        crc
    }

    /**
     * Calculate CRC-32 checksum.
     */
    fun crc32(data: ByteArray): Int {
        var crc = -1
        for (byte in data) {
            val index = (crc xor byte.toInt()) and 0xFF
            crc = (crc ushr 8) xor CRC32_TABLE[index]
        }
        return crc.inv()
    }

    /**
     * Calculate CRC-32 checksum for string.
     */
    fun crc32(string: String): Int = crc32(string.toByteArray(Charsets.UTF_8))

    /**
     * Calculate Adler-32 checksum.
     */
    fun adler32(data: ByteArray): Int {
        val mod = 65521
        var a = 1
        var b = 0

        for (byte in data) {
            a = (a + (byte.toInt() and 0xFF)) % mod
            b = (b + a) % mod
        }

        return (b shl 16) or a
    }

    /**
     * Calculate Adler-32 checksum for string.
     */
    fun adler32(string: String): Int = adler32(string.toByteArray(Charsets.UTF_8))

    /**
     * Calculate FNV-1a 32-bit hash.
     */
    fun fnv1a32(data: ByteArray): Int {
        var hash = 0x811c9dc5.toInt()
        for (byte in data) {
            hash = hash xor (byte.toInt() and 0xFF)
            hash = (hash * 0x01000193)
        }
        return hash
    }

    /**
     * Calculate FNV-1a 32-bit hash for string.
     */
    fun fnv1a32(string: String): Int = fnv1a32(string.toByteArray(Charsets.UTF_8))

    /**
     * Calculate FNV-1a 64-bit hash.
     */
    fun fnv1a64(data: ByteArray): Long {
        var hash = 0xcbf29ce484222325UL.toLong()
        for (byte in data) {
            hash = hash xor (byte.toLong() and 0xFF)
            hash = (hash * 0x00000100000001B3L)
        }
        return hash
    }

    /**
     * Calculate FNV-1a 64-bit hash for string.
     */
    fun fnv1a64(string: String): Long = fnv1a64(string.toByteArray(Charsets.UTF_8))

    /**
     * Validate using Luhn algorithm (credit cards, etc.).
     */
    fun luhnValidate(digits: String): Boolean {
        val chars = digits.filter { it.isDigit() }
        if (chars.isEmpty()) return false

        var sum = 0
        var alternate = false

        for (char in chars.reversed()) {
            val digit = char.digitToInt()
            if (alternate) {
                val doubled = digit * 2
                sum += if (doubled > 9) doubled - 9 else doubled
            } else {
                sum += digit
            }
            alternate = !alternate
        }

        return sum % 10 == 0
    }

    /**
     * Calculate Luhn check digit.
     */
    fun luhnCheckDigit(digits: String): Int? {
        val chars = digits.filter { it.isDigit() }
        if (chars.isEmpty()) return null

        var sum = 0
        var alternate = true

        for (char in chars.reversed()) {
            val digit = char.digitToInt()
            if (alternate) {
                val doubled = digit * 2
                sum += if (doubled > 9) doubled - 9 else doubled
            } else {
                sum += digit
            }
            alternate = !alternate
        }

        return (10 - (sum % 10)) % 10
    }

    /**
     * Validate ISBN-10.
     */
    fun isbn10Validate(isbn: String): Boolean {
        val chars = isbn.filter { it.isDigit() || it == 'X' || it == 'x' }
        if (chars.length != 10) return false

        var sum = 0
        for ((index, char) in chars.withIndex()) {
            val value = when {
                char == 'X' || char == 'x' -> {
                    if (index != 9) return false
                    10
                }
                else -> char.digitToInt()
            }
            sum += value * (10 - index)
        }

        return sum % 11 == 0
    }

    /**
     * Validate ISBN-13.
     */
    fun isbn13Validate(isbn: String): Boolean {
        val chars = isbn.filter { it.isDigit() }
        if (chars.length != 13) return false

        var sum = 0
        for ((index, char) in chars.withIndex()) {
            val digit = char.digitToInt()
            sum += digit * (if (index % 2 == 0) 1 else 3)
        }

        return sum % 10 == 0
    }

    /**
     * Calculate simple 8-bit checksum.
     */
    fun checksum8(data: ByteArray): Byte {
        var sum: Byte = 0
        for (byte in data) {
            sum = (sum + byte).toByte()
        }
        return sum
    }

    /**
     * Calculate simple 16-bit checksum.
     */
    fun checksum16(data: ByteArray): Short {
        var sum: Short = 0
        for (byte in data) {
            sum = (sum + (byte.toInt() and 0xFF)).toShort()
        }
        return sum
    }

    /**
     * Calculate XOR checksum.
     */
    fun xorChecksum(data: ByteArray): Byte {
        var result: Byte = 0
        for (byte in data) {
            result = (result.toInt() xor byte.toInt()).toByte()
        }
        return result
    }

    /**
     * Calculate Fletcher-16 checksum.
     */
    fun fletcher16(data: ByteArray): Short {
        var sum1 = 0
        var sum2 = 0

        for (byte in data) {
            sum1 = (sum1 + (byte.toInt() and 0xFF)) % 255
            sum2 = (sum2 + sum1) % 255
        }

        return ((sum2 shl 8) or sum1).toShort()
    }

    /**
     * Calculate Fletcher-32 checksum.
     */
    fun fletcher32(data: ByteArray): Int {
        var sum1 = 0
        var sum2 = 0
        var index = 0

        while (index + 1 < data.size) {
            val word = (data[index].toInt() and 0xFF) or ((data[index + 1].toInt() and 0xFF) shl 8)
            sum1 = (sum1 + word) % 65535
            sum2 = (sum2 + sum1) % 65535
            index += 2
        }

        if (index < data.size) {
            sum1 = (sum1 + (data[index].toInt() and 0xFF)) % 65535
            sum2 = (sum2 + sum1) % 65535
        }

        return (sum2 shl 16) or sum1
    }
}

File: .UNSAFE-RUST-DELETED-20260125/safe_checksum.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

//! Safe checksum and hash operations.
//!
//! Provides CRC, hash, and integrity verification functions
//! with overflow protection.

/// CRC-32 checksum (IEEE polynomial).
pub fn crc32(data: &[u8]) -> u32 {
    const TABLE: [u32; 256] = generate_crc32_table();
    let mut crc = 0xFFFFFFFF_u32;
    for byte in data {
        let index = ((crc ^ (*byte as u32)) & 0xFF) as usize;
        crc = (crc >> 8) ^ TABLE[index];
    }
    !crc
}

const fn generate_crc32_table() -> [u32; 256] {
    let mut table = [0u32; 256];
    let mut i = 0;
    while i < 256 {
        let mut crc = i as u32;
        let mut j = 0;
        while j < 8 {
            if crc & 1 != 0 {
                crc = (crc >> 1) ^ 0xEDB88320;
            } else {
                crc >>= 1;
            }
            j += 1;
        }
        table[i] = crc;
        i += 1;
    }
    table
}

/// Adler-32 checksum.
pub fn adler32(data: &[u8]) -> u32 {
    const MOD_ADLER: u32 = 65521;
    let mut a: u32 = 1;
    let mut b: u32 = 0;

    for byte in data {
        a = (a + *byte as u32) % MOD_ADLER;
        b = (b + a) % MOD_ADLER;
    }

    (b << 16) | a
}

/// Fletcher-16 checksum.
pub fn fletcher16(data: &[u8]) -> u16 {
    let mut sum1: u16 = 0;
    let mut sum2: u16 = 0;

    for byte in data {
        sum1 = (sum1 + *byte as u16) % 255;
        sum2 = (sum2 + sum1) % 255;
    }

    (sum2 << 8) | sum1
}

/// Simple FNV-1a hash (64-bit).
pub fn fnv1a_64(data: &[u8]) -> u64 {
    const FNV_OFFSET: u64 = 0xcbf29ce484222325;
    const FNV_PRIME: u64 = 0x100000001b3;

    let mut hash = FNV_OFFSET;
    for byte in data {
        hash ^= *byte as u64;
        hash = hash.wrapping_mul(FNV_PRIME);
    }
    hash
}

/// Simple FNV-1a hash (32-bit).
pub fn fnv1a_32(data: &[u8]) -> u32 {
    const FNV_OFFSET: u32 = 0x811c9dc5;
    const FNV_PRIME: u32 = 0x01000193;

    let mut hash = FNV_OFFSET;
    for byte in data {
        hash ^= *byte as u32;
        hash = hash.wrapping_mul(FNV_PRIME);
    }
    hash
}

/// DJB2 hash algorithm.
pub fn djb2(data: &[u8]) -> u32 {
    let mut hash: u32 = 5381;
    for byte in data {
        hash = hash.wrapping_mul(33).wrapping_add(*byte as u32);
    }
    hash
}

/// SDBM hash algorithm.
pub fn sdbm(data: &[u8]) -> u32 {
    let mut hash: u32 = 0;
    for byte in data {
        hash = (*byte as u32)
            .wrapping_add(hash.wrapping_shl(6))
            .wrapping_add(hash.wrapping_shl(16))
            .wrapping_sub(hash);
    }
    hash
}

/// Luhn algorithm for credit card validation.
pub fn luhn_check(digits: &str) -> bool {
    let chars: Vec<char> = digits.chars().filter(|c| c.is_ascii_digit()).collect();
    if chars.is_empty() {
        return false;
    }

    let mut sum = 0u32;
    let mut double = false;

    for c in chars.iter().rev() {
        let mut digit = c.to_digit(10).unwrap();
        if double {
            digit *= 2;
            if digit > 9 {
                digit -= 9;
            }
        }
        sum += digit;
        double = !double;
    }

    sum % 10 == 0
}

/// Generate Luhn check digit.
pub fn luhn_digit(digits: &str) -> Option<char> {
    let chars: Vec<char> = digits.chars().filter(|c| c.is_ascii_digit()).collect();
    if chars.is_empty() {
        return None;
    }

    let mut sum = 0u32;
    let mut double = true;

    for c in chars.iter().rev() {
        let mut digit = c.to_digit(10).unwrap();
        if double {
            digit *= 2;
            if digit > 9 {
                digit -= 9;
            }
        }
        sum += digit;
        double = !double;
    }

    let check = (10 - (sum % 10)) % 10;
    char::from_digit(check, 10)
}

/// Verify data integrity with CRC-32.
pub fn verify_crc32(data: &[u8], expected: u32) -> bool {
    crc32(data) == expected
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_crc32() {
        let data = b"123456789";
        assert_eq!(crc32(data), 0xCBF43926);
    }

    #[test]
    fn test_adler32() {
        let data = b"Wikipedia";
        assert_eq!(adler32(data), 0x11E60398);
    }

    #[test]
    fn test_fnv1a() {
        let data = b"hello";
        assert_ne!(fnv1a_64(data), 0);
        assert_ne!(fnv1a_32(data), 0);
    }

    #[test]
    fn test_luhn() {
        // Valid Visa test number
        assert!(luhn_check("4532015112830366"));
        // Invalid
        assert!(!luhn_check("1234567890123456"));
    }

    #[test]
    fn test_luhn_digit() {
        assert_eq!(luhn_digit("453201511283036"), Some('6'));
    }
}
